samza-azure/src/main/java/org/apache/samza/AzureJobCoordinator.java

```diff
@@ -0,0 +1,483 @@
+            Thread.interrupted();
+          }
+          LOG.info("Checking for barrier state on the blob again...");
+          blobBarrierState = leaderBlob.getBarrierState();
+        }
+      }
+    }
+  }
+
+  /**
+   * Called when the JC detects that the barrier has completed by checking the barrier state on the blob.
+   * @param nextJMVersion The new job model version after rebalancing.
+   */
+  private void onNewJobModelConfirmed(final String nextJMVersion) {
+    LOG.info("pid=" + processorId + "new version " + nextJMVersion + " of the job model got confirmed");
+
+    // Delete previous value
+    if (table.getEntity(currentJMVersion.get(), processorId) != null) {
+      table.deleteProcessorEntity(currentJMVersion.get(), processorId);
+    }
+    if (table.getEntity(INITIAL_STATE, processorId) != null) {
+      table.deleteProcessorEntity(INITIAL_STATE, processorId);
+    }
+
+    //Start heartbeating to new entry only when barrier reached.
+    //Changing the current job model version enables that since we are heartbeating to a row identified by the current job model version.
+    currentJMVersion.getAndSet(nextJMVersion);
+
+    //Start the container with the new model
+    if (coordinatorListener != null) {
+      coordinatorListener.onNewJobModel(processorId, jobModel);
+    }
+  }
+
+  private String createProcessorId(Config config) {
+    // TODO: This check to be removed after 0.13+
+    ApplicationConfig appConfig = new ApplicationConfig(config);
+    if (appConfig.getProcessorId() != null) {
+      return appConfig.getProcessorId();
+    } else if (StringUtils.isNotBlank(appConfig.getAppProcessorIdGeneratorClass())) {
+      ProcessorIdGenerator idGenerator =
+          ClassLoaderHelper.fromClassName(appConfig.getAppProcessorIdGeneratorClass(), ProcessorIdGenerator.class);
+      return idGenerator.generateProcessorId(config);
+    } else {
+      throw new ConfigException(String
+          .format("Expected either %s or %s to be configured", ApplicationConfig.PROCESSOR_ID,
+              ApplicationConfig.APP_PROCESSOR_ID_GENERATOR_CLASS));
+    }
+  }
+
+  public class AzureLeaderElectorListener implements LeaderElectorListener {
+    /**
+     * Keep renewing the lease and do the required tasks as a leader.
+     */
+    @Override
+    public void onBecomingLeader() {
+      // Update table to denote that it is a leader.
+      table.updateIsLeader(currentJMVersion.get(), processorId, true);
+
+      // Schedule a task to renew the lease after a fixed time interval
+      LOG.info("Starting scheduler to keep renewing lease held by the leader.");
+      renewLease.scheduleTask();
+
+      doOnProcessorChange(new ArrayList<>());
+
+      // Start scheduler to check for change in list of live processors
+      LOG.info("Starting scheduler to check for change in list of live processors in the system.");
+      liveness.setStateChangeListener(createLivenessListener(liveness.getLiveProcessors()));
```

Comment:
If there is failure in renewing the lease, I am updating the isLeader value of that processor to be false and resigning the leadership. Also it will shutdown all the schedulers that the leader is supposed to run. This will indicate to the other processors that the current leader died and someone else will become the leader and start all these tasks again.